### PR TITLE
Support custom spreading/interpolation kernels

### LINKED ISSUE #10: Expose `interp_*d` and `spread_*d` functions for more general usage?
Hello again! Again, awesome package. I was wondering the following: would you be open to exposing API that would support more general usage of your `interp_*d` and `spread_*d` functions? I'd like to use these for spreading of short-ranged gaussians and other kernels, without using the full nufft pipeline.

An oversimplification of what I'm thinking is changing the following (here e.g. only for `spread_1d`):

```python
def spread_1d(
    x: jax.Array,
    c: jax.Array,
    nf: int,
    kernel_params: KernelParams,  # Specifically for nufft kernel
) -> jax.Array:
    ...
```

To something like this:

```python
def spread_1d(
    x: jax.Array,
    c: jax.Array,
    nf: int,
    kernel_fn: Callable[[Array], [Array]],  # Supports generic `f(z) = f_val`
    kernel_with_derivative_fn: Callable[[Array], [Array, Array]],  # And `f(z) = f_val, df_dz_val`
    nspread: int,
) -> jax.Array:
    ...
```

I know a request like this could be beyond the scope of the package, but I'm curious to hear what you think. I'd be happy to put together a PR.

### PR DESCRIPTION
**Pull request by geoffroyO**
Closes #10.

Exposes `spread_*` / `interp_*` for use with a user-supplied kernel instead of the built-in ES kernel — e.g. spreading short-range Gaussians without the full NUFFT pipeline.

**API**
- New `Kernel(nspread, phi, phi_and_dphi=None)` — `phi` is the kernel value, `phi_and_dphi` its optional analytic derivative (autodiff fallback otherwise, used for gradients w.r.t. coordinates).
- `spread_*` / `interp_*` now accept either `KernelParams` (ES, unchanged) or a `Kernel`.

**GPU**
Custom kernels run through the same fused Pallas kernels — `phi` is threaded into the Triton kernel as a static closure. The ES path takes the `phi=None` branch, byte-identical to before.

### CODE REVIEW

**Comment by michael-0brien** [comment]
Looks like we still need to add the Kernel class to the type hint in the `nufftax.core.spread_*d` and `nufftax.core.interp_*d` functions?

**Comment by michael-0brien** [comment]
Also, I think the frozen dataclass may break JAX transforms w.r.t. the kernel parameters, which in general we do support. If you’d consider taking on [`equinox`](https://github.com/patrick-kidger/equinox) as a dependency, which has lots of helpful JAX features, this can be fixed by making the `Kernel` an `equinox.Module`.

Alternatively, the `Kernel` could also be a named tuple and a helper function could be used to return the relevant `value` and `value_and_grad` functions!

**Comment by geoffroyO** [comment]
#12

**Comment by geoffroyO** [comment]
@michael-0brien If this looks good to you, I’ll merge and deploy it.

@Jammy2211 Just wanted to give you a heads-up since I know you’re using nufftax in PyAutoLens and this PR introduces some substantial changes.

**Comment by michael-0brien** [comment]
This looks great! Thanks so much for merging these in, I’m excited to try things out.
